File: depth_first_search/1145_binary_tree_coloring_game/1145_binary_tree_coloring_game.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

        
class Solution:
    def btreeGameWinningMove(self, root: TreeNode, n: int, x: int) -> bool:
        self.red_left, self.red_right = 0, 0
        
        def dfs(root):
            if not root:
                return 0
            left = dfs(root.left)
            right = dfs(root.right)
            if root.val == x:
                self.red_left = left
                self.red_right = right
            return left + right + 1
        
        dfs(root)
        parent = n - self.red_left - self.red_right - 1
        judge = [parent, self.red_left, self.red_right]
        return any([j > n // 2 for j in judge])

File: depth_first_search/1145_binary_tree_coloring_game/1145_binary_tree_coloring_game.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func btreeGameWinningMove(root *TreeNode, n int, x int) bool {

	var (
		target     *TreeNode
		count      func(root *TreeNode) int
		findTarget func(root *TreeNode)
	)

	count = func(root *TreeNode) int {
		if root == nil || root == target {
			return 0
		}
		return 1 + count(root.Left) + count(root.Right)
	}

	findTarget = func(root *TreeNode) {
		if root != nil {
			if root.Val == x {
				target = root
				return
			}
			findTarget(root.Left)
			findTarget(root.Right)
		}
	}

	findTarget(root)

	return count(root) > n/2 || count(target.Left) > n/2 || count(target.Right) > n/2
}
